src/audio/mux/mux_generic.c

```diff
@@ -26,8 +26,8 @@
  * \param[in] offset Offset in source buffer.
  * \param[in] mask Routing bitmask for calculating output sample.
  */
-UT_STATIC inline int32_t calc_sample_s16le(const struct audio_stream *source,
-					   uint32_t offset, uint8_t mask)
```

Comment:
About this commit, "inline" is a hint for the optimizer and changes semantics for the linker as its main thing. So the reasoning in the commit message looks unsound to me. What did you experience without doing this?

Reply:
Interesting @paulstelian97 can you explain what linker difference `inline` makes?

For gcc for instance https://gcc.gnu.org/onlinedocs/gcc/Inline.html
> Therefore, a non-static inline function is always compiled on its own in the usual fashion.

Does the linker care about static functions and how?

PS: as far as actual inlining is concerned, I've never seen gcc care much about the hint.

Reply:
inline (without static) implies weak linkage, or the possibility of multiple instances of the same symbol to exist without an error. The linker will for each use select any weak symbol or THE strong symbol when actually doing the symbol resolution. It's an unusual semantic that will always make me raise questions.

Reply:
So, `inline` is a compiler-specific and C version specific nightmare. This patch removes it - what's not to like? :-)

PS: I agree the commit message is too vague.

Reply:
I'm using the same (and experimental) branch than @finikorg and I spotted this warning, this is probably what he saw too:

```
modules/audio/sof/zephyr/../src/include/sof/audio/audio_stream.h:161:2: warning:
'audio_stream_wrap' is static but used in inline function 'calc_sample_s16le' which is not static
  audio_stream_wrap(buffer, (char *)(ptr) + ((idx) * (sample_size)))
```

Reply:
Without this patch we have inline without static and i does not build with current Zephyr CFLAGS

Reply:
```
zephyr-sdk/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/9.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(mux_generic.c.obj):(.literal.demux_s32le+0x0): undefined reference to `calc_sample_s32le'
zephyr-sdk/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/9.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(mux_generic.c.obj):(.literal.demux_s24le+0x0): undefined reference to `calc_sample_s24le'
zephyr-sdk/xtensa/intel_apl_adsp/xtensa-zephyr-elf/bin/../lib/gcc/xtensa-zephyr-elf/9.2.0/../../../../xtensa-zephyr-elf/bin/ld: modules/sof/libmodules_sof.a(mux_generic.c.obj):(.literal.demux_s16le+0x0): undefined reference to `calc_sample_s16le'
```
etc

Reply:
There is always the combination "static inline" which is what I recommend for functions defined in headers. It would generate it once per compilation unit and allow inlining at caller side, without creating anything visible to the linker.